Capture stopwatch results

tortmayr opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See SECURITY.md at the root of this repository, to learn how to report
vulnerabilities.
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->
Introduce `MeasurementResults` that capture a stopwatch execution in a serializable format. 
Add `onMeasurementResult` event emitter to stopwatch class and introduce optional  caching of results.
Add a metrics contribution to `@theia/metrics` that collects all measurement results of frontend and backend processes (when running with debug log level).

The new metrics can be picked up by the  performance measurement step of the [e2e framework](https://github.com/eclipse-theia/theia-e2e-test-suite/pull/9).

Contributed on behalf of STMicroelectronics

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->
Start the example Theia application in browser in debug mode (`yarn browser start:debug`). Go to `localhost:3000/metrics`. The stopwatch results for the backend and frontend should be part of the dataset.
Reload the theia application, then reload `localhost:3000/metrics`. It should now contain an additional dataset for the second frontend instance.
#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

tsmaeder commented on packages/core/src/common/performance/measurement.ts:
```diff
@@ -101,4 +101,30 @@ export interface MeasurementOptions {
+     * Flag to indicate whether the stopwatch should cache measurement results for later retrieval.
+     * For example the cache can be used to retrieve measurements which were taken during startup before a listener had a chance to register.
+     */
+    cacheResults?: boolean
```
In IT, the word "cache" usually refers to a short-term, incomplete set of items, often size-limited,  where some items may be replaced according to some algorithm. I suspect this is not what's happening here. If we want to store a full trace of results, we should call this field `storeResults` and rename related variables accordingly.

tsmaeder commented on packages/core/src/common/performance/stopwatch.ts:
```diff
@@ -50,10 +51,20 @@ export abstract class Stopwatch {
     @inject(ILogger)
     protected readonly logger: ILogger;
 
+    protected cachedResults: MeasurementResult[] = [];
```
`storedMeasurements`

tsmaeder commented on packages/core/src/common/performance/stopwatch.ts:
```diff
@@ -50,10 +51,20 @@ export abstract class Stopwatch {
 
+    protected cachedResults: MeasurementResult[] = [];
+
+    protected onMeasurementResultEmitter = new Emitter<MeasurementResult>();
```
According to our [guidelines](https://github.com/eclipse-theia/theia/blob/master/doc/coding-guidelines.md#event_names), the name of this event should be `onDidAddMeasurementResult`.

tsmaeder commented on packages/core/src/common/performance/stopwatch.ts:
```diff
@@ -91,14 +102,17 @@ export abstract class Stopwatch {
     }
 
-    protected createMeasurement(name: string, measurement: () => number, options?: MeasurementOptions): Measurement {
+    protected createMeasurement(name: string, measurement: () => { startTime: number, duration: number }, options?: MeasurementOptions): Measurement {
```
I would prefer `measure` as a name, since it's a function.

tsmaeder commented on packages/core/src/common/performance/stopwatch.ts:
```diff
@@ -112,6 +126,19 @@ export abstract class Stopwatch {
         return result;
     }
 
+    protected handleCompletedMeasurement(name: string, startTime: number, elapsed: number, options: LogOptions): void {
```
Why not keep this method private or inline it? I don't see the point of overriding it in a subclass. It just adds gratuitous API surface.

tsmaeder commented on packages/core/src/common/performance/stopwatch.ts:
```diff
@@ -154,4 +181,8 @@ export abstract class Stopwatch {
         this.logger.log(level, `${whatWasMeasured}: ${elapsed.toFixed(1)} ms [${timeFromStart}]`, ...(options.arguments ?? []));
     }
 
+    getCachedResults(): MeasurementResult[] {
```
Why not use a property accessor returning a readonly array? It prevent unnecessary copying.

tsmaeder commented on packages/metrics/src/browser/index.ts:
```diff
@@ -0,0 +1,17 @@
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
+// *****************************************************************************
+
+export * from './metrics-frontend-application-contribution';
```
Isn't the point of the contribution to be used in the module? Why does it need to be exported from index.ts?

tsmaeder commented on packages/metrics/src/node/measurement-metrics-contribution.ts:
```diff
@@ -0,0 +1,75 @@
+    protected logLevelCli: LogLevelCliContribution;
+
+    protected metrics = '';
+    protected frontendCounter = new Map<string, string>();
```
`frontendCounters`

tsmaeder commented:
Here's the text form localhost:3000/metrics after starting a front end:
```
# HELP process_cpu_user_seconds_total Total user CPU time spent in seconds. # TYPE process_cpu_user_seconds_total counter process_cpu_user_seconds_total 0.516 1693146828441 # HELP process_cpu_system_seconds_total Total system CPU time spent in seconds. # TYPE process_cpu_system_seconds_total counter process_cpu_system_seconds_total 0.078 1693146828441 # HELP process_cpu_seconds_total Total user and system CPU time spent in seconds. # TYPE process_cpu_seconds_total counter process_cpu_seconds_total 0.594 1693146828441 # HELP process_start_time_seconds Start time of the process since unix epoch in seconds. # TYPE process_start_time_seconds gauge process_start_time_seconds 1693146482 # HELP process_resident_memory_bytes Resident memory size in bytes. # TYPE process_resident_memory_bytes gauge process_resident_memory_bytes 85241856 1693146828441 # HELP nodejs_eventloop_lag_seconds Lag of event loop in seconds. # TYPE nodejs_eventloop_lag_seconds gauge nodejs_eventloop_lag_seconds 0.0001915 1693146828441 # HELP nodejs_active_handles_total Number of active handles. # TYPE nodejs_active_handles_total gauge nodejs_active_handles_total 23 1693146828441 # HELP nodejs_active_requests_total Number of active requests. # TYPE nodejs_active_requests_total gauge nodejs_active_requests_total 0 1693146828441 # HELP nodejs_heap_size_total_bytes Process heap size from node.js in bytes. # TYPE nodejs_heap_size_total_bytes gauge nodejs_heap_size_total_bytes 49008640 1693146828441 # HELP nodejs_heap_size_used_bytes Process heap size used from node.js in bytes. # TYPE nodejs_heap_size_used_bytes gauge nodejs_heap_size_used_bytes 46841368 1693146828441 # HELP nodejs_external_memory_bytes Nodejs external memory size in bytes. # TYPE nodejs_external_memory_bytes gauge nodejs_external_memory_bytes 19389753 1693146828441 # HELP nodejs_heap_space_size_total_bytes Process heap space size total from node.js in bytes. # TYPE nodejs_heap_space_size_total_bytes gauge nodejs_heap_space_size_total_bytes{space="read_only"} 176128 1693146828441 nodejs_heap_space_size_total_bytes{space="old"} 24633344 1693146828441 nodejs_heap_space_size_total_bytes{space="code"} 2203648 1693146828441 nodejs_heap_space_size_total_bytes{space="map"} 2105344 1693146828441 nodejs_heap_space_size_total_bytes{space="large_object"} 18841600 1693146828441 nodejs_heap_space_size_total_bytes{space="code_large_object"} 0 1693146828441 nodejs_heap_space_size_total_bytes{space="new_large_object"} 0 1693146828441 nodejs_heap_space_size_total_bytes{space="new"} 1048576 1693146828441 # HELP nodejs_heap_space_size_used_bytes Process heap space size used from node.js in bytes. # TYPE nodejs_heap_space_size_used_bytes gauge nodejs_heap_space_size_used_bytes{space="read_only"} 170944 1693146828441 nodejs_heap_space_size_used_bytes{space="old"} 23872880 1693146828441 nodejs_heap_space_size_used_bytes{space="code"} 2030624 1693146828441 nodejs_heap_space_size_used_bytes{space="map"} 1304784 1693146828441 nodejs_heap_space_size_used_bytes{space="large_object"} 18754816 1693146828441 nodejs_heap_space_size_used_bytes{space="code_large_object"} 0 1693146828441 nodejs_heap_space_size_used_bytes{space="new_large_object"} 0 1693146828441 nodejs_heap_space_size_used_bytes{space="new"} 710912 1693146828441 # HELP nodejs_heap_space_size_available_bytes Process heap space size available from node.js in bytes. # TYPE nodejs_heap_space_size_available_bytes gauge nodejs_heap_space_size_available_bytes{space="read_only"} 0 1693146828441 nodejs_heap_space_size_available_bytes{space="old"} 293408 1693146828441 nodejs_heap_space_size_available_bytes{space="code"} 25568 1693146828441 nodejs_heap_space_size_available_bytes{space="map"} 761144 1693146828441 nodejs_heap_space_size_available_bytes{space="large_object"} 0 1693146828441 nodejs_heap_space_size_available_bytes{space="code_large_object"} 0 1693146828441 nodejs_heap_space_size_available_bytes{space="new_large_object"} 1031072 1693146828441 nodejs_heap_space_size_available_bytes{space="new"} 320160 1693146828441 # HELP nodejs_version_info Node.js version info. # TYPE nodejs_version_info gauge nodejs_version_info{version="v16.17.0",major="16",minor="17",patch="0"} 1 # HELP theia_extension_version Theia extension version info. # TYPE theia_extension_version gauge theia_extension_version{extension="theia_electron",version="1.40.0"} 1 theia_extension_version{extension="theia_core",version="1.40.0"} 1 theia_extension_version{extension="theia_variable_resolver",version="1.40.0"} 1 theia_extension_version{extension="theia_editor",version="1.40.0"} 1 theia_extension_version{extension="theia_filesystem",version="1.40.0"} 1 theia_extension_version{extension="theia_process",version="1.40.0"} 1 theia_extension_version{extension="theia_workspace",version="1.40.0"} 1 theia_extension_version{extension="theia_file_search",version="1.40.0"} 1 theia_extension_version{extension="theia_markers",version="1.40.0"} 1 theia_extension_version{extension="theia_outline_view",version="1.40.0"} 1 theia_extension_version{extension="theia_monaco",version="1.40.0"} 1 theia_extension_version{extension="theia_output",version="1.40.0"} 1 theia_extension_version{extension="theia_navigator",version="1.40.0"} 1 theia_extension_version{extension="theia_search_in_workspace",version="1.40.0"} 1 theia_extension_version{extension="theia_userstorage",version="1.40.0"} 1 theia_extension_version{extension="theia_toolbar",version="1.40.0"} 1 theia_extension_version{extension="theia_bulk_edit",version="1.40.0"} 1 theia_extension_version{extension="theia_callhierarchy",version="1.40.0"} 1 theia_extension_version{extension="theia_console",version="1.40.0"} 1 theia_extension_version{extension="theia_terminal",version="1.40.0"} 1 theia_extension_version{extension="theia_task",version="1.40.0"} 1 theia_extension_version{extension="theia_debug",version="1.40.0"} 1 theia_extension_version{extension="theia_editor_preview",version="1.40.0"} 1 theia_extension_version{extension="theia_messages",version="1.40.0"} 1 theia_extension_version{extension="theia_preferences",version="1.40.0"} 1 theia_extension_version{extension="theia_scm",version="1.40.0"} 1 theia_extension_version{extension="theia_timeline",version="1.40.0"} 1 theia_extension_version{extension="theia_typehierarchy",version="1.40.0"} 1 theia_extension_version{extension="theia_plugin_ext",version="1.40.0"} 1 theia_extension_version{extension="theia_plugin_ext_vscode",version="1.40.0"} 1 theia_extension_version{extension="theia_vsx_registry",version="1.40.0"} 1 theia_extension_version{extension="theia_api_samples",version="1.40.0"} 1 theia_extension_version{extension="theia_keymaps",version="1.40.0"} 1 theia_extension_version{extension="theia_getting_started",version="1.40.0"} 1 theia_extension_version{extension="theia_scm_extra",version="1.40.0"} 1 theia_extension_version{extension="theia_git",version="1.40.0"} 1 theia_extension_version{extension="theia_memory_inspector",version="1.40.0"} 1 theia_extension_version{extension="theia_metrics",version="1.40.0"} 1 theia_extension_version{extension="theia_mini_browser",version="1.40.0"} 1 theia_extension_version{extension="theia_plugin_dev",version="1.40.0"} 1 theia_extension_version{extension="theia_plugin_metrics",version="1.40.0"} 1 theia_extension_version{extension="theia_preview",version="1.40.0"} 1 theia_extension_version{extension="theia_property_view",version="1.40.0"} 1 
```

Beyond the confusing format, I'm not seeing any outputs the look stopwatch-related.

tortmayr commented:
Thanks for the review @tsmaeder. I have rebased the PR and addresses your comments.

> Beyond the confusing format, I'm not seeing any outputs the look stopwatch-related.

The test instructions were incomplete, my bad. Currently the measurement metrics is only active if the application is started in debug mode (`yarn start browser:debug`), I have updated the instructions accordingly.

tsmaeder reviewed (approved):
Thanks for addressing the comments, I can see the metrics now.

tsmaeder commented:
I'm approving based on the assumption that the `/metrics` endpoint is not to be consumed by humans. If it _is_ for humans, we need to fix the formatting.
